### PR TITLE
Rspec tests for book model

### PR DESCRIPTION
**Pull request by alanlamaswolox**
## Summary
I installed the rspec gem and the test for the book model

## Trello
https://trello.com/c/6l4YgSIH/22-12-crear-tests-para-el-modelo-de-book

### CODE REVIEW

File: spec/models/book_spec.rb [hunk 1]
```diff
@@ -0,0 +1,92 @@
+require 'rails_helper'
+
+describe Book, type: :model do
+  subject(:book) { described_class.new(gender: gender, author: author, image: image, title: title, editor: editor, year: year) }
+
+  let(:gender) { Faker::Book.genre }
```
Comment (fedeest) [inline spec/models/book_spec.rb]: This is the first time I notice this. Our bad, should've told you earlier that it should be genre istead of gender.

File: spec/models/book_spec.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
+require 'rails_helper'
+
+describe Book, type: :model do
+  subject(:book) { described_class.new(gender: gender, author: author, image: image, title: title, editor: editor, year: year) }
```
Comment (fedeest) [inline spec/models/book_spec.rb]: Few things:

1. We create the models we use with a factory using the gem FactoryBot.
2. The models are created inside a `let`
3. To validate validations of models, we use shoulda-matchers `https://github.com/thoughtbot/shoulda-matchers`

So, the scheme should be as follows:
```
let(:book) { create(:book) }

(...)

it { is_expected.to validate_presence_of(:author) }
```


And in the factory file for book (`spec/factories/book.rb`):

```
FactoryBot.define do
  factory :book do
    author { Faker::Book.author }
    (...)
  end
end
```

File: spec/spec_helper.rb [hunk 3]
```diff
@@ -0,0 +1,94 @@
+# This file was generated by the `rails generate rspec:install` command. Conventionally, all
```
Comment (fedeest) [inline spec/spec_helper.rb:2]: Add `require 'factory_bot'` here

File: spec/spec_helper.rb [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+# Given that it is always loaded, you are encouraged to keep this file as
+# light-weight as possible. Requiring heavyweight dependencies from this file
+# will add to the boot time of your test suite on EVERY test run, even for an
+# individual file that may not need all of that loaded. Instead, consider making
+# a separate helper file that requires the additional dependencies and performs
+# the additional setup, and require it from the spec files that actually need
+# it.
+#
+# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
+RSpec.configure do |config|
```
Comment (fedeest) [inline spec/spec_helper.rb:17]: Within this, include `config.include FactoryBot::Syntax::Methods` so you can call FactoryBot directly as I said in me previous comment without doing `FactoryBot.create` all the time and just do `create`.

File: .rubocop.yml [hunk 5]
```diff
@@ -2,4 +2,7 @@ Documentation:
   Enabled: false
 
 LineLength:
-  Max: 99
+  Max: 150
```
Comment (nicozar95) [inline .rubocop.yml]: Don't change de max line length of the whole rubocop, you can change it on the spec files only like:
```
Metrics/LineLength:
  Exclude:
    - spec/**/*
```

But i dont suggest you do this, check my review on the spec

File: .rubocop.yml [hunk 6]
```diff
@@ -2,4 +2,7 @@ Documentation:
...
+
+BlockLength:
+  Max: 100
```
Comment (nicozar95) [inline .rubocop.yml]: The same as before, i dont care for big clases if you are testing:
```
Metrics/BlockLength:
  Exclude:
    - spec/**/*
```

File: spec/models/user_spec.rb [hunk 7]
```diff
@@ -0,0 +1,44 @@
+require 'rails_helper'
```
Comment (fedeest) [inline spec/models/user_spec.rb]: The card didn't specify to add tests for user. Beware from now on to comply with the scope of the card you are working on.

File: app/models/book.rb [hunk 8]
```diff
@@ -1,3 +1,3 @@
 class Book < ApplicationRecord
-  validates %i[gender author image title editor year], presence: true
+  validates :gender, :author, :image, :title, :editor, :year, presence: true
```
Comment (nicozar95) [inline app/models/book.rb]: Why did you change the array of syms?

File: spec/models/user_spec.rb [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+describe User, type: :model do
```
Comment (fedeest) [inline spec/models/user_spec.rb]: Apply the same corrections I mentioned on the previous comment here aswell.

**Review by fedeest — CHANGES_REQUESTED** [review]

File: spec/models/book_spec.rb (hunk 2 above)
Comment (nicozar95) [inline spec/models/book_spec.rb]: Dont change the line length because of this, make it more easy to read by adding new lines

File: spec/models/book_spec.rb [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+  let(:image) { 'ImagenFalsa123.gif' }
+  let(:title) { Faker::Book.title }
+  let(:editor) { Faker::Book.publisher }
+  let(:year) { '2006' }
+
+  it do
+    is_expected.to be_valid
+  end
+
+  describe '#Create' do
```
Comment (nicozar95) [inline spec/models/book_spec.rb]: `#create` dont capitalize it

File: spec/models/book_spec.rb [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    context 'when the user is created' do
+      it 'also creates its random gender' do
```
Comment (nicozar95) [inline spec/models/book_spec.rb]: remove `also`

Comment (nicozar95) [reply spec/models/book_spec.rb]: Because, this is a way of documentation and when the test run it would said: `when the user is created also creates its random gender`

Comment (nicozar95) [reply spec/models/book_spec.rb]: I will no test the happy test that it works, i would only test if you are creating it with null values

File: spec/models/book_spec.rb [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+  subject(:book) { described_class.new(gender: gender, author: author, image: image, title: title, editor: editor, year: year) }
+
+  let(:gender) { Faker::Book.genre }
+  let(:author) { Faker::Book.author }
+  let(:image) { 'ImagenFalsa123.gif' }
+  let(:title) { Faker::Book.title }
+  let(:editor) { Faker::Book.publisher }
+  let(:year) { '2006' }
+
+  it do
```
Comment (nicozar95) [inline spec/models/book_spec.rb]: Can you make this into a one line block?

File: spec/models/book_spec.rb [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+      it 'also creates its random editor' do
+        expect(book.editor).to be_present
+      end
+
+      it 'also creates its year, 2006' do
+        expect(book.year).to be_present
+      end
+    end
+
+    context 'When author is nil' do
```
Comment (nicozar95) [inline spec/models/book_spec.rb]: Dont capitalize the description on a context

File: spec/models/user_spec.rb [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+  let(:last_name)             { Faker::Name.last_name }
+  let(:email)                 { Faker::Internet.email }
+  let(:password)              { 'password' }
+  let(:password_confirmation) { 'password' }
+
+  it do
+    is_expected.to be_valid
+  end
+
+  describe '#Generate users email' do
```
Comment (nicozar95) [inline spec/models/user_spec.rb]: `#create`

File: spec/models/user_spec.rb [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    context 'When the user is created' do
+      it 'Generates a random user email' do
```
Comment (nicozar95) [inline spec/models/user_spec.rb]: Remove this kind of test and only test if you can created with an invalid value, it should fail in that case

File: spec/models/user_spec.rb (hunk 14 above)
Comment (nicozar95) [reply spec/models/user_spec.rb]: Test by method of a model, if you are calling `#create` i assume there is a method name `created` in the model

File: spec/models/user_spec.rb [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+        expect(user.email).to be_present
+      end
+    end
+  end
+
+  describe '#create' do
+    context 'When the first_name is nil' do
+      let(:first_name) { nil }
+
+      it do
```
Comment (nicozar95) [inline spec/models/user_spec.rb]: Dont use 1 line blocks with `do end`. Use `{}`

**Review by nicozar95 — CHANGES_REQUESTED** [review]
There are a couple of changes to make

File: db/migrate/20180914124755_create_books.rb [hunk 17]
```diff
@@ -2,11 +2,11 @@ class CreateBooks < ActiveRecord::Migration[5.1]
   def change
     create_table :books do |t|
 
-      t.string :gender, null:false
```
Comment (nicozar95) [inline db/migrate/20180914124755_create_books.rb:5]: Beware of changing the migration directly, You should have created a new migration to change de DB column name, if you change the last migration directly then you have to do a rollback and run the old migration again which may cause troubles in the DB. Now the DB is short but in future PR please make a new migration to change the table

File: spec/factories/book.rb [hunk 18]
```diff
@@ -0,0 +1,10 @@
+FactoryBot.define do
+  factory :book do
+    author    { Faker::Book.author }
+    genre     { Faker::Book.genre }
+    image     { 'ImagenFalsa123.gif' }
```
Comment (nicozar95) [inline spec/factories/book.rb]: Use Faker for this

File: spec/factories/book.rb [hunk 19]
```diff
@@ -0,0 +1,10 @@
...
+    title     { Faker::Book.title }
+    publisher { Faker::Book.publisher }
+    year      { '2006' }
```
Comment (nicozar95) [inline spec/factories/book.rb]: Use Faker for this

**Review by nicozar95 — CHANGES_REQUESTED** [review]
Almost done, just some minor changes. The migration changes doesn't bother me that much but keep it in mind for future PRs

File: db/migrate/20180914124755_create_books.rb (hunk 17 above)
Comment (alanlamaswolox) [reply db/migrate/20180914124755_create_books.rb:5]: Oh yes, i dont see that. In the future  i'll try not to make the same mistake

**Review by nicozar95 — APPROVED** [review]
👍 Great Work!!!

**Review by fedeest — APPROVED** [review]
